src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelModuleResolutionFunction.java

```diff
@@ -366,8 +366,16 @@ static BazelModuleResolutionValue createValue(
     // Calculate a unique name for each used extension id.
     BiMap<String, ModuleExtensionId> extensionUniqueNames = HashBiMap.create();
     for (ModuleExtensionId id : extensionUsagesById.rowKeySet()) {
-      String bestName =
-          id.getBzlFileLabel().getRepository().getName() + "~" + id.getExtensionName();
+      // Ensure that the resulting extension name (and thus the repository names derived from it) do
+      // not start with a tilde.
+      String repoName = id.getBzlFileLabel().getRepository().getName();
+      String nonEmptyRepoPart;
+      if (repoName.isEmpty()) {
+        nonEmptyRepoPart = "_main";
+      } else {
+        nonEmptyRepoPart = repoName;
+      }
+      String bestName = nonEmptyRepoPart + "~" + id.getExtensionName();
```

Comment:
Nit: Maybe we also check if `id.getExtensionName()` is empty? It probably should be done early: https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java;l=407

Reply:
It's not strictly necessary since a path ending but not starting with a tilde should not require escaping. 

Wouldn't an empty `extensionName` already lead to an error since it needs to name a Starlark export in the file and these cannot have empty names?

Reply:
Ah, that make sense.